File: app/src/main/java/s15666/pjwstk/pamobmi/ui/bmi/BmiParamWatcher.java
package s15666.pjwstk.pamobmi.ui.bmi;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import java.util.Locale;

import s15666.pjwstk.pamobmi.bmi.BmiViewModel;
import s15666.pjwstk.pamobmi.bmi.calculator.BmiCalculator;

class BmiParamWatcher implements TextWatcher {

    private BmiViewModel model;

    private EditText weightField;
    private EditText heightField;

    private BmiResultUpdater resultUpdater;
    private BmiCalculator calculator;

    BmiParamWatcher(BmiViewModel model, EditText weightField, EditText heightField,
                    BmiResultUpdater resultUpdater, BmiCalculator calculator) {
        this.model = model;
        this.weightField = weightField;
        this.heightField = heightField;
        this.resultUpdater = resultUpdater;
        this.calculator = calculator;

        loadModelData();
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        // not implemented
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        // not implemented
    }

    private void loadModelData() {
        Double weight = model.getWeight().getValue();
        Double height = model.getHeight().getValue();

        // TODO get currently used locale instead of hardcoding US
        if(weight != null) {
            weightField.setText(String.format(Locale.US, "%.2f", model.getWeight().getValue()));
        }

        if(height != null) {
            heightField.setText(String.format(Locale.US, "%.2f", model.getHeight().getValue()));
        }
    }

    @Override
    public void afterTextChanged(Editable s) {
        double weight;
        double height;

        try {
            weight = Double.valueOf(weightField.getText().toString());
            height = Double.valueOf(heightField.getText().toString());

            model.setWeight(weight);
            model.setHeight(height);
        } catch (NumberFormatException e) {
            return;
        }

        resultUpdater.update(calculator.calculate(weight, height));
    }

    void setCalculator(BmiCalculator calculator) {
        this.calculator = calculator;
        afterTextChanged(null);
    }
}

File: app/src/main/java/s15666/pjwstk/pamobmi/ui/bmi/BmiParamWatcher.kt
package s15666.pjwstk.pamobmi.ui.bmi

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import android.widget.TextView
import s15666.pjwstk.pamobmi.bmi.BmiViewModel
import s15666.pjwstk.pamobmi.bmi.calculator.BmiCalculator
import java.util.*
import java.lang.Double as JDouble

internal class BmiParamWatcher(
        private val model: BmiViewModel,
        private val weightField: EditText,
        private val heightField: EditText,
        private val resultUpdater: BmiResultUpdater,
        private var calculator: BmiCalculator) : TextWatcher {

    private val invalid = -1.0

    init {
        loadModelData()
    }

    private fun loadModelData() {
        val weight = model.weight.value ?: invalid
        val height = model.height.value ?: invalid

        weightField.setText(String.format(Locale.US, getFormat(weight), weight))
        heightField.setText(String.format(Locale.US, getFormat(height), height))
    }

    private fun getFormat(value: Double): String = if(value == invalid) "" else "%.2f"

    // not implemented
    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

    override fun afterTextChanged(s: Editable?) {
        updateModelField(model::setWeight, weightField)
        updateModelField(model::setHeight, heightField)

        if(!model.isValid.value!!)
            resultUpdater.clear()
        else
            resultUpdater.update(calculator.calculate(model.weight.value!!, model.height.value!!))
    }

    private fun updateModelField(field: (v: Double) -> Unit, text: TextView) {
        field(try { JDouble.valueOf(text.text.toString()) } catch(e: NumberFormatException) { invalid })
    }

    fun setCalculator(calculator: BmiCalculator) {
        this.calculator = calculator
        afterTextChanged(null)
    }

}